client/infinispan-key-value-store-hotrod/src/main/java/org/infinispan/api/reactive/client/impl/KeyValueStoreImpl.java

```diff
@@ -0,0 +1,138 @@
+package org.infinispan.api.reactive.client.impl;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+import org.infinispan.api.client.listener.ClientKeyValueStoreListener;
+import org.infinispan.api.reactive.client.impl.listener.ClientListenerImpl;
+import org.infinispan.api.exception.InfinispanException;
+import org.infinispan.api.reactive.KeyValueStore;
+import org.infinispan.api.reactive.query.QueryRequest;
+import org.infinispan.api.reactive.listener.KeyValueStoreListener;
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.client.hotrod.Search;
+import org.infinispan.query.api.continuous.ContinuousQuery;
+import org.infinispan.query.dsl.Query;
+import org.infinispan.query.dsl.QueryFactory;
+import org.reactivestreams.Publisher;
+
+import io.reactivex.Flowable;
+import io.reactivex.processors.UnicastProcessor;
+
+/**
+ * Implements the {@link KeyValueStore} interface
+ *
+ * @author Katia Aresti, karesti@redhat.com
+ * @since 10.0
+ */
+public class KeyValueStoreImpl<K, V> implements KeyValueStore<K, V> {
+   protected final RemoteCache<K, V> cache;
+   protected RemoteCache<K, V> cacheReturnValues;
+
+   public KeyValueStoreImpl(RemoteCache<K, V> cache, RemoteCache<K, V> cacheReturnValues) {
+      this.cache = cache;
+      this.cacheReturnValues = cacheReturnValues;
+   }
+
+   @Override
+   public CompletionStage<V> get(K key) {
+      return cache.getAsync(key);
+   }
+
+   @Override
+   public CompletionStage<Void> insert(K key, V value) {
+      // We don't return the value here
+      return cache.putIfAbsentAsync(key, value).thenApply(v -> null);
+   }
+
+   @Override
+   public CompletionStage<Void> save(K key, V value) {
+      // We don't return the value here
+      return cache.putAsync(key, value).thenApply(v -> null);
+   }
+
+   @Override
+   public CompletionStage<Void> delete(K key) {
+      return cache.removeAsync(key).thenApply(v -> null);
+   }
+
+   @Override
+   public Publisher<K> keys() {
+      UnicastProcessor<K> processor = UnicastProcessor.create();
+      cache.keySet().iterator().forEachRemaining(k -> {
+         processor.onNext(k);
+      });
+      return processor;
+   }
+
+   @Override
+   public Publisher<? extends Map.Entry<K, V>> entries() {
+      UnicastProcessor<Map.Entry<K, V>> processor = UnicastProcessor.create();
+      cache.entrySet().iterator().forEachRemaining(e -> {
+         processor.onNext(e);
+      });
+      return processor;
+   }
+
+   @Override
+   public CompletionStage<Void> saveMany(Publisher<Map.Entry<K, V>> pairs) {
+      return CompletableFuture.runAsync(() -> {
+         Flowable<Map.Entry<K, V>> entryFlowable = Flowable.fromPublisher(pairs);
+         entryFlowable.subscribe(e -> cache.putAsync(e.getKey(), e.getValue()),
```

Comment:
I'm not sure how this works, but isn't the saveMany() completion stage completed before all entries are stored?

Reply:
yes; this is very weird and not ok

Reply:
I think this should return a Publisher of write results intead

Reply:
reconsidering signature